### PR TITLE
EQL: Add field resolution and verification

### PR DESCRIPTION
**Pull request by costin**
Add basic field resolution inside the Analyzer and a basic Verifier to
check for any unresolved fields

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Analyzer.java [hunk 1]
```diff
@@ -42,4 +50,56 @@ private LogicalPlan verify(LogicalPlan plan) {
...
+                        }
+                        return named;
+                    }
+                }
+                return e;
+            });
+        }
+    }
+
+    abstract static class AnalyzeRule<SubPlan extends LogicalPlan> extends Rule<SubPlan, LogicalPlan> {
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Analyzer.java]: There will be a separate PR to externalize this to QL so it can be shared across SQL and EQL.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -11,20 +11,16 @@
...
 import org.elasticsearch.xpack.ql.type.DataTypes;
 
-import static java.util.Collections.emptyMap;
-
 public abstract class LogicalPlanBuilder extends ExpressionBuilder {
 
     // TODO: these need to be made configurable
-    private static final String EVENT_TYPE = "event.category";
-    private static final EsIndex esIndex = new EsIndex("<not-specified>", emptyMap());
+    private static final String EVENT_TYPE = "event_type";
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java]: I changed this since the data sample from existing EQL documentation uses `event_type`. If the default field should be changed, that is best addressed separately:
- figure out what's the impact on backwards compatibility (ECS schema) and existing data
- update datasets
- update code (last).

**Comment by costin** [comment]
Follow-up issue to beef the verifier tests further here: https://github.com/elastic/elasticsearch/issues/51873

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java (hunk 2 above)
Comment (aleksmaus) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java]: was assuming that the data will be in ECS format
```{
    "user": {
        "group": {}
    },
    "host": {
        "os": {
            "platform": "windows",
            "name": "Windows"
        },
        "ip": "127.0.0.1",
        "hostname": "localhost",
        "name": "localhost"
    },
    "event": {
        "module": "endgame",
        "dataset": "esensor",
        "action": "already_running",
        "category": "process",
        "kind": "event"
    }, 
```
thus the event.category for EVENT_TYPE

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -43,6 +39,6 @@ public LogicalPlan visitEventQuery(EqlBaseParser.EventQueryContext ctx) {
 
         }
 
-        return new Filter(source(ctx), new EsRelation(Source.EMPTY, esIndex, false), condition);
+        return new Filter(source(ctx), new UnresolvedRelation(Source.EMPTY, null, "", false, ""), condition);
```
Comment (rw-access) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java]: 👍

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/logical/UnresolvedRelation.java [hunk 4]
```diff
@@ -89,11 +89,11 @@ public boolean equals(Object obj) {
         }
 
         UnresolvedRelation other = (UnresolvedRelation) obj;
-        return source().equals(other.source())
-            && table.equals(other.table)
+        return Objects.equals(source(), other.source())
```
Comment (rw-access) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/logical/UnresolvedRelation.java]: why does `source()` need to be checked here, but isn't in some of the other hashCode and equals functions?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/AnalysisUtils.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        if (matches.isEmpty()) {
+            return null;
+        }
+
+        if (matches.size() == 1) {
+            return handleSpecialFields(u, matches.get(0), allowCompound);
+        }
+
+        return u.withUnresolvedMessage(
+                "Reference [" + u.qualifiedName() + "] is ambiguous (to disambiguate use quotes or qualifiers); matches any of "
```
Comment (rw-access) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/AnalysisUtils.java]: We may have to come back and clarify "quotes" in this message for #51443

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/logical/UnresolvedRelation.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/logical/UnresolvedRelation.java]: It's an inconsistency - the source isn't relevant. Likely for UnresolvedRelation was added since a table or index declaration is not common and thus, its source was considered relevant.
The change above was added to avoid a NPE caused by a null (vs an empty) location.

**Review by rw-access — APPROVED** [review]

**Review by aleksmaus — APPROVED** [review]

**Review by astefan — APPROVED** [review]
LGTM
